test/src/test/integration/git_protocol.rs

```diff
@@ -0,0 +1,384 @@
+// Copyright © 2021 The Radicle Link Contributors
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{
+    collections::BTreeSet,
+    io,
+    path::Path,
+    process::ExitStatus,
+    sync::{atomic::AtomicBool, Arc},
+};
+
+use bstr::ByteSlice as _;
+use futures::{AsyncReadExt as _, TryFutureExt as _};
+use radicle_link_git_protocol::{
+    fetch,
+    packwriter,
+    upload_pack,
+    ObjectId,
+    PackWriter,
+    Ref,
+    WantRef,
+};
+use tempfile::{tempdir, TempDir};
+
+fn upstream() -> TempDir {
+    let tmp = tempdir().unwrap();
+
+    let repo = git2::Repository::init_bare(&tmp).unwrap();
+    let auth = git2::Signature::now("apollo", "apollo@cree.de").unwrap();
+
+    let tree = {
+        let empty = repo.treebuilder(None).unwrap();
+        let oid = empty.write().unwrap();
```

Comment:
Actually all of work done here except for `Repository::init_bare(…)` and `.set_namespace()` can already be done in `gitoxide`. Maybe once all is supported I could contribute a PR which oxidizes the test setup.

Reply:
That'd be awesome.